django/db/migrations/state.py

```diff
@@ -188,6 +193,89 @@ def _reload(self, related_models):
         # Render all models
         self.apps.render_multiple(states_to_be_rendered)
 
+    def resolve_fields_and_relations(self):
+        for model_state in self.models.values():
+            for field_name, field in model_state.fields:
+                field.name = field_name
+                field.model = "{}.{}".format(model_state.app_label, model_state.name)
+                if field.remote_field and field.remote_field.model:
+                    self.fill_remote_field_name(field)
+
+        self.relations = defaultdict(lambda: defaultdict(list))
+        concretes, proxies = self._get_concrete_models_mapping_and_proxy_models()
+
+        real_apps = set(self.real_apps)
+        for model_key in concretes:
+            model_state = self.models[model_key]
+            for field_name, field in model_state.fields:
+                remote_field = field.remote_field
+                if remote_field:
+                    remote_model_key = resolve_model_key(*model_key, remote_field.model)
+                    if remote_model_key[0] not in real_apps and remote_model_key in concretes:
+                        remote_model_key = concretes[remote_model_key]
+                    self.relations[remote_model_key][model_key].append((field_name, field))
+
+                    through = getattr(remote_field, 'through', None)
+                    if through and not model_state.options.get('auto_created'):
+                        through_model_key = resolve_model_key(*model_key, through)
+                        if through_model_key[0] not in real_apps:
+                            through_model_key = concretes[through_model_key]
+                        self.relations[through_model_key][model_key].append((field_name, field))
+        for model_key in proxies:
+            self.relations[model_key] = self.relations[concretes[model_key]]
+
+    def fill_remote_field_name(self, field):
+        if self.is_swappable_model(field.remote_field.model):
+            field.remote_field.field_name = "id"
+        else:
+            try:
+                model_state = self.get_concrete_model_state(field.remote_field.model)
+                base = model_state.bases[0]
+                if not isinstance(base, str) and base.__name__ == "Model":
```

Comment:
This seem a bit arbitrary, is it the original logic? I'd type check against `base` instead.

Reply:
This comes from the default bases we attribute on ModelState: https://github.com/django/django/blob/master/django/db/migrations/state.py#L366 
But you are right that it seems that it isn't needed actually.

`if not isinstance(base, str):` seems to be enough. I think I had troubles making tests pass when I didn't add catching the KeyError.